Nhse d34 nhskv.i33 token

martinsumner opened the pull request:
Add support for a token manager and token session process to act as a requestor for tokens, and warp riak_client calls within a session.

This is then used to be provide a stronger conditional PUT API to allow for check-and-set operations that will behave reliably when the cluster is healthy, as well as in simple failure scenarios and operational changes.  The CAS will fall-back to eventual consistency in extreme failure scenarios.

martinsumner committed:
Initial support for OTP26

martinsumner committed:
Remove tracers

martinsumner committed:
Update rebar.config

martinsumner committed:
Update rebar.config

martinsumner committed:
Token Manager

Start a token manager on each node.  The token manager will on request register a session against a token, and remove the registration when the session terminates.  If another request is received for the same token, it will be queued until the previous session releases the token (or the queued session dies).

There is a session server that can be started to run a session (which request a token from the local token_manager on startup), and will allow for the session to be used by making calls to riak_client functions.  Each call will renew the session.  After token timeout, the session will terminate.

martinsumner committed:
Add strong conditional check to PB API

Allow, by configuration, for the conditional check to be strengthened to use the new token-based mechanism

martinsumner committed:
Reset state on PB connection after completion

martinsumner committed:
Add downstream recording of state

Means that if a topology change occurs without node failure, sibling protection is still maintained.

martinsumner committed:
Fix uniquenes of node list

Also catch erpc errors when initial connecting to node, and have back-off and retry.  This smooths handling of the case where a node is suddenly killed.

martinsumner committed:
Refactor riak_kv_pb_object

To make it simpler and clearer to remove session token

martinsumner committed:
Move strong conditional check to FSM (#34)

If the strong conditional check is at the API - becomes much harder to make the check identically and reliably in the HTTP API.

Webmachine has lots of potential exit points outside of our control - so it is hard to reliably release a session on exit.

martinsumner committed:
Add support to HTTP API

martinsumner committed:
Conditional PUT to require GET not HEAD

Change the GET run as part of a conditional PUT, so as not ot require a full object fetch.

martinsumner committed:
Add configuration

martinsumner committed:
Add profile function

To help in riak_test

martinsumner committed:
Add profiler

martinsumner committed:
Merge branch 'nhse-d34-otp26' into nhse-d34-nhskv.i30-profiler

martinsumner committed:
Type fix

martinsumner committed:
Add to extending list of defaults

martinsumner committed:
Merge branch 'nhse-d34-nhskv.i30-profiler' into nhse-d34-nhskv.i33-token

martinsumner committed:
Initial write-up of riak_kv_token_manager

Also allows for the async update to verifications that an upstream node has granted a token

ThomasArts commented on src/riak_kv_token_session.erl:
```diff
@@ -0,0 +1,467 @@
+    TargetNodeCount =
+        case NVal of
+            1 -> 1;
+            TNC when TNC > 1 -> TNC -2
```
I wonder about this hard coded 2.
Compared to the documentation in `token_manager` 
```
%% token.  The next two nodes in the preflist (that are UP), are responsible
%% for verifying that the grant can be made (i.e. to confirm that they have not
%% granted a token while the head node was recently unavailable). 
```
Isn't the NVal used in `token_session` actually Target Nval and should then the TargetNodeCount not be `NVal-1` for the actual NVal, not the TargetNVal  - 2?

martinsumner replied:
Changed as part of renaming the configuration options

ThomasArts commented on src/riak_kv_token_session.erl:
```diff
@@ -0,0 +1,467 @@
+
+session_request_retry(_TokenID, _NVal, 0, _TokenTO, _Retry, _Attempts) ->
+    {false, none};
+session_request_retry(TokenID, NVal, RequestTO, TokenTO, Retry, Retry) ->
```
1. I find counting down more logic than counting up. That is, start with Retry and then reduce until zero
2. or until 1, because now you do Retries + 1 attempts.

ThomasArts commented on src/riak_kv_token_session.erl:
```diff
@@ -0,0 +1,467 @@
+    ok;
+session_release(SessionReference) ->
+    {N, P, ID} = decode_session_reference(SessionReference),
+    case node() of
```
Note that after restaring a node, the Pid `P` may have been gone, which is fine... but in very rare circumstances, it could also be the case that that node actually has this Pid, but that that process is running some totally different code. Very unlikely, but....

What happens then is that an arbitrary Process may be getting a gen_server:call from `sessio_local_use`.  Since that has timeout ìnfinity` and since it may be a process that just discards the message, you may have a deadlock there, waiting for every to use the session that no longer exists.

We must make an argument to explain that this cannot possibly happen or change it in such a way that this won't happen... Somehow would be nice if we could identify the process we use by its "SessionRef". For example if a session would be a registerd process and we would use the token as its name or so. But that would fill the atom table, so no good.
We could ask the manager if the Pid still belongs to the token? But probably too many indirections.

martinsumner committed:
Refactor to use monitoring

Discussion of the reliability of process monitoring, including the monitoring of processes on remote nodes, led to a new design which simplifies the message passing required by relying instead on monitoring.

A request should now be refused if there exists an currently active grant, and the preflist has changed since that grant was activated.  In all other situations grants should be made (either immediately or after being queued awaiting a release).

martinsumner committed:
Change to config parameters

With aim to improve clarity

martinsumner committed:
Remove duplicate

martinsumner committed:
Rename (again), make messages async

The naming has been changed  following review to try and clarify meaning.

The downstream_check was previously sync - but two token_managers could attempt to check in each other at the same time (for different tokens) - resulting in deadlock.

These messages are now async not sync.

martinsumner committed:
Clarify erpc errors

ThomasArts commented on priv/riak_kv.schema:
```diff
@@ -540,6 +540,74 @@
+%% required in this mode).
+%%
+%% With basic_consensus, tokens can still be granted in a wide range of failure
+%% scenarios, but with a risk of duplicate grants, in aprticular should a
```
```suggestion
%% scenarios, but with a risk of duplicate grants, in particular should a
```

ThomasArts commented on src/riak_kv_sup.erl:
```diff
@@ -98,10 +98,14 @@ init([]) ->
 
+    TokenManager = {riak_kv_token_manager,
+                    {riak_kv_token_manager, start_link, []},
+                    permanent, 3000, worker, [riak_kv_token_manager]},
```
A short shutdown timeout here...
This means that when you shutdown a node, all ongoing token_sessions will be released, but ongoing work may possibly continue a bit more, since other processes are slighty slower to stop working.
This seems like the right approach to me.

ThomasArts commented on src/riak_kv_token_manager.erl:
```diff
@@ -0,0 +1,610 @@
+%% 
+%% riak_kv_token_session process may request a grant.  The session process
+%% should monnitor the token_manager to which it makes a request, and fail with
+%% the fialure of that token_manager.  A request may be granted immediately, or
```
```suggestion
%% the failure of that token_manager.  A request may be granted immediately, or
```

ThomasArts commented on src/riak_kv_token_manager.erl:
```diff
@@ -0,0 +1,610 @@
+                    }
+            end;
+        _ ->
+            %% No assocition is assumed to be a token manager
```
```suggestion
            %% No association is assumed to be a token manager
```

ThomasArts replied:
If this "assume is too weak", then one could use a tag when starting the monitor and know what kind of process is down. Not sure you really need to know...
`monitor(process, Pid, [{tag, {'DOWN', Kind, Node}}]),`
which results in down messages of the form:
 `{{'DOWN', Kind, Node}, _Ref, process, Pid, _} ->`

martinsumner replied:
Now checks Ref/Pid is that of a monitored remote manager

martinsumner committed:
Use token_manager to control downstream messages

The token_manager will no longer monitor remote upstream sessions.  Add a backup GC process, whereby if a request is blocked, there is a check of the existence in the remote manager of an association with the session.

The only thing that must be reliable, is therefore the monitoring of local sessions by managers.

Before using a local session, an association check is now made, that the local token manager has an active association for that session.

martinsumner committed:
Add GC process

Very hard to test the GC process - hard to create a PID which doesn't trigger down on die.

There was a discrepancy in sloppy_quorum between PB/WM APIs for the conditional check.  This is resolved.

martinsumner committed:
Avoid re-fetching objects (#36)

Once a HEAD has been converted to a GET, disqualify it from the find_bestobject calculation.

As GETs occur after HEADs in the GET_FSM, it should by default be the case when allow_mult=true that VCget >= VChead.

This is not the case with lww=true, wher eobjects may regress in VC terms.  However as lww=true, it is destined to prefer the GET which came after the HEAD.

The code as-is could cause a HEAD to continuously require re-fetching, and lead to requests timing out.  This is demonstrated in the riak_test associated with the conditional PUT token implementation

martinsumner committed:
Merge branch 'nhse-d34-nhskv.i33-token' of https://github.com/nhs-riak/riak_kv into nhse-d34-nhskv.i33-token

martinsumner committed:
Extend eunit test of token_manager

martinsumner committed:
Try and keep github formatter happy

ThomasArts commented on src/riak_kv_token_manager.erl:
```diff
@@ -613,8 +631,31 @@ monitor_manager(Manager, MonitoredManagers) ->
+    case is_process_alive(Session) of
+        true ->
+            check_active(Rest, Mgr);
+        false ->
```
Well... possibly better to just monitor this dead pid, because then we don't have to know exactly what format the monitor uses.... 
Not really important but it clarifies that we have a process that we lost the monitor for and now we repair that,

martinsumner replied:
Changed as suggested

martinsumner committed:
Use monitor rather than spoof 'DOWN'

Change following review

martinsumner committed:
Use uniq code in OTP 24

lists:uniq introduced in OTP25

martinsumner committed:
Update comments for perceived clarity

martinsumner commented:
Looking at the results of a full performance test, the overhead of performing basic_consensus for a proportion of PUTs is not large

![Throughput Comparison - CAS](https://github.com/nhs-riak/riak_kv/assets/1628897/a1301445-23b2-4630-9572-09961ad4b69f)

martinsumner committed:
Merge branch 'nhse-develop-3.4' into nhse-d34-nhskv.i33-token

martinsumner committed:
Remove double-definition on merge

martinsumner committed:
Merge branch 'openriak-3.4' into nhse-d34-nhskv.i33-token
